File: app/src/main/java/com/aresid/happyapp/signup/FirstNameFragment.java
package com.aresid.happyapp.signup;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.aresid.happyapp.R;
import com.aresid.happyapp.keys.Keys;
import com.aresid.happyapp.utils.Utils;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

/**
 * Created on: 23/04/2020
 * For Project: HappyApp
 * Author: René Spies
 * Copyright: © 2020 ARES ID
 */

public class FirstNameFragment
		extends Fragment
		implements View.OnClickListener {
	
	private static final String            TAG = "FirstNameFragment";
	private              NavController     mNavController;
	private              TextInputEditText mFirstNameField;
	private              TextInputLayout   mFirstNameFieldLayout;
	
	public FirstNameFragment() {
		
		Log.d(TAG, "FirstNameFragment: called");
		
		// Required empty constructor
		
	}
	
	@Nullable
	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		
		Log.d(TAG, "onCreateView: called");
		
		// Inflate the layout
		return inflater.inflate(R.layout.fragment_first_name, container, false);
		
	}
	
	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		
		Log.d(TAG, "onViewCreated: called");
		
		super.onViewCreated(view, savedInstanceState);
		
		// Define the NavController object
		mNavController = Navigation.findNavController(view);
		
		// Set onClickListener
		view.findViewById(R.id.next_button).setOnClickListener(this);
		
		// Define first name field TextInputEditText from layout
		mFirstNameField = view.findViewById(R.id.first_name_field);
		
		// Define first name field layout TextInputLayout from layout
		mFirstNameFieldLayout = view.findViewById(R.id.first_name_field_layout);
		
	}
	
	@Override
	public void onClick(View v) {
		
		Log.d(TAG, "onClick: called");
		
		if (v.getId() == R.id.next_button) {
			
			showFamilyNameFragment();
			
		}
		
	}
	
	/**
	 * Bundles the first name and navigates to the FamilyNameFragment passing the Bundle.
	 */
	private void showFamilyNameFragment() {
		
		Log.d(TAG, "showFamilyNameFragment: called");
		
		// Show error if no first name has been typed in
		if (mFirstNameField.length() == 0) {
			
			// No first name detected
			
			// Show error
			mFirstNameFieldLayout.setError(getString(R.string.error_you_forgot_me));
			
			return;
			
		}
		
		// Set the first name fields layout error to null proactively
		mFirstNameFieldLayout.setError(null);
		
		// Define a new Bundle object
		Bundle arguments = new Bundle();
		
		// Put the first name into the bundle
		arguments.putString(Keys.BundleKeys.KEY_FIRST_NAME, Utils.getString(mFirstNameField.getText()));
		
		// Navigate to the FamilyNameFragment and pass the bundle
		mNavController.navigate(R.id.to_familyNameFragment, arguments);
		
	}
	
}

File: app/src/main/java/com/aresid/happyapp/signup/form/FirstNameFragment.kt
package com.aresid.happyapp.signup.form

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.NavHostFragment.findNavController
import com.aresid.happyapp.R
import com.aresid.happyapp.databinding.FragmentFirstNameBinding
import com.aresid.happyapp.signup.EmailSignupViewModel
import timber.log.Timber

/**
 * Created on: 23/04/2020
 * For Project: HappyApp
 * Author: René Spies
 * Copyright: © 2020 ARES ID
 */
class FirstNameFragment: Fragment() {
	
	// Binding for the layout
	private lateinit var binding: FragmentFirstNameBinding
	
	// Corresponding ViewModel
	private lateinit var firstNameViewModel: FirstNameViewModel
	
	// EmailSignupViewModel needed to pass the firstName
	private lateinit var emailSignupViewModel: EmailSignupViewModel
	
	override fun onCreateView(
		inflater: LayoutInflater,
		container: ViewGroup?,
		savedInstanceState: Bundle?
	): View? {
		
		Timber.d("onCreateView: called")
		
		// Define the binding
		binding = FragmentFirstNameBinding.inflate(
			inflater,
			container,
			false
		)
		
		// Define the ViewModel
		firstNameViewModel = ViewModelProvider(this).get(FirstNameViewModel::class.java)
		
		// Tell the binding about the ViewModel
		binding.viewModel = firstNameViewModel
		
		// Observe the firstNameEmpty LiveData and show feedback if true
		firstNameViewModel.firstNameEmpty.observe(viewLifecycleOwner,
		                                          Observer { isError ->
			
			                                          // If isError, show feedback to the user
			                                          if (isError) {
				
				                                          // Show an error message on the firstNameFieldLayout
				                                          binding.firstNameFieldLayout.error = getString(R.string.error_you_forgot_me)
				
			                                          }
			
			                                          // Else, reset the error
			                                          else {
				
				                                          binding.firstNameFieldLayout.error = null
				
			                                          }
			
		                                          })
		
		// Observe the firstNameOk LiveData, notify the emailSignupViewModel about the firstName and navigate to the FamilyNameFragment
		firstNameViewModel.firstNameOk.observe(viewLifecycleOwner,
		                                       Observer { isOk ->
			
			                                       // If isOk, notify and navigate
			                                       if (isOk) {
				
				                                       // Put the firstName into the SignupFormData singleton object
				                                       SignupFormData.getInstance().firstName = firstNameViewModel.firstName.value!!
				
				                                       // Navigate to the FamilyNameFragment
				                                       navigateToFamilyNameFragment()
				
				                                       // Reset the firstNameOk LiveData
				                                       firstNameViewModel.navigatedAndNotified()
				
			                                       }
			
		                                       })
		
		// Return the inflated layout
		return binding.root
		
	}
	
	/**
	 * Navigates to the FamilyNameFragment using the NavController.
	 */
	private fun navigateToFamilyNameFragment() {
		
		Timber.d("navigateToFamilyNameFragment: called")
		
		// Navigate to the FamilyNameFragment
		findNavController(this).navigate(FirstNameFragmentDirections.toFamilyNameFragment())
		
	}
	
}